playwright/_impl/_transport.py

```diff
@@ -97,3 +116,43 @@ def send(self, message: Dict) -> None:
         self._output.write(
             len(data).to_bytes(4, byteorder="little", signed=False) + data
         )
+
+
+class WebSocketTransport(AsyncIOEventEmitter, Transport):
+    def __init__(self, ws_endpoint: str, timeout: float) -> None:
+        super().__init__()
+        self._stopped = False
+        self.ws_endpoint = ws_endpoint
+        self.timeout = timeout
+        self._loop: asyncio.AbstractEventLoop
+
+    def stop(self) -> None:
+        self._stopped = True
+        self._loop.create_task(self._connection.close())
+
+    async def wait_until_stopped(self) -> None:
+        await self._connection.wait_closed()
+
+    async def run(self) -> None:
+        self._loop = asyncio.get_running_loop()
+
+        options = {}
+        if self.timeout is not None:
+            options["close_timeout"] = self.timeout / 1000
+            options["ping_timeout"] = self.timeout / 1000
+        self._connection = await websockets.connect(self.ws_endpoint, **options)
+
+        while not self._stopped:
+            try:
+                message = await self._connection.recv()
+                obj = json.loads(message)
+                self.on_message(obj)
+            except websockets.exceptions.ConnectionClosed:
+                self.emit("close")
+                break
+            except Exception as exc:
+                print("got unhandled exception", exc)
```

Comment:
Raise Playwright `Error`

Reply:
Will probably remove that completely, was more for debugging. That's why its a draft PR, it was mostly not yet ready for review.